http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/HandlerRemovingChannelPool.java

```diff
@@ -70,10 +70,17 @@ public void close() {
 
     private void removePerRequestHandlers(Channel channel) {
         channel.attr(IN_USE).set(false);
-        removeIfExists(channel.pipeline(),
-                       HttpStreamsClientHandler.class,
-                       ResponseHandler.class,
-                       ReadTimeoutHandler.class,
-                       WriteTimeoutHandler.class);
+
+        // Only remove per request handler if the channel is registered
+        // or open since DefaultChannelPipeline would remove handlers if
+        // channel is closed and unregistered
+        // See DefaultChannelPipeline.java#L1403
```

Comment:
Why doesn't removeIfExists prevent the issue? Is the name on the method wrong?

Reply:
The name is not technically wrong and what it does is to first check if the handler is null and removing it if not null.

```
            if (pipeline.get(handler) != null) {
                pipeline.remove(handler);
            }
```

The issue is that when a channel is being closed and unregistered, another thread will remove handlers on its pipeline, so there is a race condition when two threads trying to remove the handlers at the same time. Basically, right after a thread goes into the if block,  another thread has removed the handler.

Reply:
With the current catch added, is this line 78 change still needed?

Reply:
It'd be nice to have line 18 to reduce the chances of the race condition because filling stacktrace for an exception is expensive.

Reply:
Are these two calls cheap?

Reply:
Yeah, should be. They are basically just getters.